### ISSUE TITLE
OpenStack LBaaS resources - need to disassociate monitor from pool before destroy

### ISSUE THREAD

**Issue by martineg** [hashicorp/terraform#1715]
Provisioning an OpenStack based stack with LBaaS resources fails on destroy.
An LBaaS monitor associated with a pool needs to be disassociated from the pool before either resource can be deleted. Currently Terraform does not disassociate the resources and destroy fails.

Given the below resources, `terraform destroy` gives an error (after a 500 error from Neutron).  Only after manually disassociating the health monitor from the pool (`neutron lb-healthmonitor-disassociate <monitor id> <pool id>`) destroy runs OK and the pool and monitor resources are destroyed as expected.

```
resource "openstack_lb_monitor_v1" "monitor_1" {
..
}

resource "openstack_lb_pool_v1" "pool_1" {
  name = "tf-pool-1"
  ..
  monitor_ids = ["${openstack_lb_monitor_v1.monitor_1.id}"]
}
```

Environment: Terraform 0.4.2, OpenStack Juno

**Comment by DanyC97** [comment]
where was this issue left?

**Comment by ggiamarchi** [comment]
@martineg I'm not sure to understand you problem. Why do you get a 500 error from neutron ? This should never happen and it is more likely to be a bug on the neutron side. I'm used to manage LBaaS using terraform with a configuration like yours and i did not encounter any problem.

**Comment by martineg** [comment]
@ggiamarchi Neutron gives me a 500 if trying to delete the pool that has health monitors associated.  The health monitor and the pool can only be deleted after first disassociating the health monitor from the pool.

We do use the A10 LBaaS driver. I see that the InUse exception caught comes from the acos-client used by that driver, so this behaviour might be specific to environments using the a10-openstack-lbaas driver.

**Comment by jtopjian** [comment]
I agree with @ggiamarchi here: a 500 means that something bad happened. If this was a matter of timing, you'd see a non-500 error code (I apologize, but it's escaping me at the moment).

I recommend trying the latest master (Or waiting until 0.6.9 is released) as there was a patch (#3926) added to assist with timing issues. If you're still seeing the 500 errors, I'm not sure what the best approach is. The a10 driver looks to be related to a vendor-based load balancer, so it'd be hard (if not impossible?) to test without the equipment (but let me know if I'm wrong). Maybe report the 500 error to a10? But either way, let us know the results and if we can help.

**Comment by desktophero** [comment]
I'm not getting feedback in the form of HTTP responses when running TF commands but experiencing similar outcomes just a TIMEOUT in trying to do the thing. 

The feedback I get form an `apply` command when an LB is in place is a timeout when attempting to apply new `user_data` to a VM. This results in a new VM instance and a new member being instantiated within the LB pool.

So far, I have to manually:
* Remove the monitor
* Remove the members
* Delete the VIP
* Delete the LB

Then I can push new user_data and get my VM(s) rebuilt. I'm trying to dig in deeper but my visibility into the OpenStack environment is purely as a tenant owner so access to logs, etc., is extremely limited.

**Comment by jtopjian** [comment]
@desktophero Do you know if the OpenStack environment is using a proprietary load balancing driver?

**Comment by desktophero** [comment]
I will dig in and circle back

**Comment by desktophero** [comment]
I need to clarify a thing and that might drive a different/new issue. I can `destroy` without a problem when executing `terraform destroy`. When I have an existing plan and need to modify a VM resource that's a member of a pool, the timeouts occur.

My friendly OpenStack admin confirmed the driver is the default haproxy neutron-lbaas-agent driver.

**Comment by jtopjian** [comment]
@desktophero Thanks for the info. I think the prior discussion was with regard to a proprietary driver.

However, knowing that you're running into this issue with the standard driver and during the plan, refresh, and update steps is very good to know. I'll look into this.

**Comment by desktophero** [comment]
Any update? I ask as we've experienced some other conditions for other resource types during an `apply` where a rebuild is needed - or at least that's the initial view into the `TIMEOUT` response.

I'm going to push for more detail but there's some view into `TIMEOUT` taking place in the `router-interface-v2` resource type during an apply. I don't have more information at this time but didn't want to lose sight of this out there.

**Comment by jtopjian** [comment]
@desktophero Coincidentally I was looking into this a few days ago.

I was able to build a "full stack" load balancer, from instances up to load balancer members. I then modified the `user_data` of the instances, causing a rebuild of them. Everything rebuilt correctly and without error.

Can you provide a sanitized Terraform configuration that is able to reproduce this issue on your end?

**Comment by desktophero** [comment]
I'll work on a sanitized TF config as well. Probably by Monday. Thanks!

**Comment by jtopjian** [comment]
@desktophero sorry for dropping the ball on this one. The change has been merged. let me know if you're still seeing an issue here.

**Comment by jtopjian** [comment]
I'm going to go ahead and close this issue as I believe it has been fixed. However, please let me know if you're still seeing this problem.
